runestone/common/js/runestonebase.js

```diff
@@ -212,6 +219,18 @@ export default class RunestoneBase {
             this.checkLocalStorage(); // just go right to local storage
             this.csresolver("local");
```

Comment:
Now that I look at this more carefully, I *think*  that the call to `this.csresolver()` is solving the same problem as adding the runestone-component-ready class and triggering: `runestone:component-ready`.   The difference is that it is resolving a promise that other objects could `await` or `.then`

*Maybe* using that mechanism is better than adding another one?  Note that the `csresolver` is pretty new and something I added recently which does not look widely used....

Reply:
Hmm, interesting. I remember seeing that in the code, but not looking at it very deeply. My thoughts:

- This is really nice. I like the idea of allowing code to `await` on a `csresolver` promise rather than the old jQuery emit logic. Are there any components which don't invoke `checkServer` in the constructor? If so, I'd suggest creating the promise in the constructor, so it can be waited on at any point.
- From a test perspective, there's no clean way to await a JavaScript promise from Python test code. There's probably a hack (set a global variable in JS then poll for it from Python/Selenium), but since there's already good support for waiting for a class to be present, I see no advantage in doing this. So, I'd suggest keeping the code which adds the `runestone-component-ready` class. In the future, I could see adding a CSS style to visually indicate when a component is fully loaded, making this useful for more than just testing.
- This promise doesn't always tell the whole story -- for example, the poll component does some additional server-side operations (loading from the `getpollresults` endpoint), so just awaiting on `csresolver` doesn't work for all components.

Reply:
I just pushed 646d3611 to address this. It uses a different promise (`this.component_ready_promise`), so that we can distinguish between `checkServer` competing (with the `this.checkServerComplete` promise) and the component being ready (which differs from `checkServer` completion for some components, such as poll).

Reply:
checkServer is not used by activecode or any of the non-gradable components.  So I think having a separate promise is good, and all components can then use their own logic to ensure that when the component is ready then the component_ready_promise can be resolved.  For those that do use checkServer we should ensure that checkServer is resolved before resolving component_ready.

Reply:
I agree that we should not use crazy hacks for testing when there is a tried and true method for us to use.  Adding a class is fine and as you say may have good benefits later.

Reply:
Sounds good. Since `checkServer()` calls `indicate_component_ready` only after the server data is loaded, we're in good shape!